### ISSUE TITLE
Running code and path completion are slow if project has folder with many files

### ISSUE THREAD

**Issue by stefanocoretta** [rstudio/rstudio#14375]
### System details

    RStudio Edition : Desktop
    RStudio Version : 2023.12.1+402
    OS Version      : macOS Sonoma
    R Version       : 4.3.3

### Steps to reproduce the problem

Create a Quarto project with a folder contains thousands of large files. (not an R package, just a Quarto proejct)

### Describe the problem in detail

When running code chunks from a qmd file and when completing paths with tab, all very slow.

### Describe the behavior you expected

Be able to let RStudio ignore folders if they are in .gitignore.

- [x] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [x] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [ ] If I am reporting an RStudio crash, I have included a [diagnostics report](https://support.posit.co/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by kevinushey** [comment]
Thanks for the bug report. Can you please provide a reproducible example? I tried the following steps, but didn't encounter any issue. First, I created a Quarto Project, and then created a large number of files with:

```
dir.create("subdir")
setwd("subdir")
paths <- sprintf("%05i.txt", 1:10000)
file.create(paths)
setwd("..")
```

Then, I created and saved a Quarto document, and tried running code in it. The IDE remained responsive for me.

**Comment by stefanocoretta** [comment]
Hi! The culprit is one of the folders in my project which is 14.5 GB, with about 400,000 mp3 files. I moved that outside the Quarto project and RStudio becomes responsive as usual. So I am pretty sure the issue is that RStudio is trying do index that folder and slows down. 

I tried with 400,000 files and RStudio slows down:

```r
dir.create("subdir")
setwd("subdir")
paths <- sprintf("%05i.txt", 1:400000)
file.create(paths)
setwd("..")
```

**Comment by olivroy** [comment]
I have already experienced this. but this seems to be more frequent on Network drive. and within a folder also with larger file. 

For example, if I try reading a path in my Downloads folder, then it takes a while before completions show.

I feel like it is just because the path completion has to wait for all files to be indexed before showing completions..

**Comment by stefanocoretta** [comment]
Yes, I think it's exactly that. That's why being able to ignore folders for the indexing is a helpful feature. :)
